### PR TITLE
add BART survival learner

### PR DESCRIPTION
**Pull request by bblodfon**
I "hacked" some stuff in this Bayesian learner, so some things could be improved. Notes:
- Inside [predict()](https://github.com/bblodfon/mlr3extralearners/blob/main/R/learner_BART_surv_bart.R#L229) I need the train data as it is transformed by `BART` using one of their internal functions. I saved that in the `self$state` [slot](https://github.com/bblodfon/mlr3extralearners/blob/main/R/learner_BART_surv_bart.R#L205). Maybe there is a better way?
- Since this a Bayesian model, it outputs not just a survival matrix (obs x times) but also posterior draws (3rd dimension) which is nice for uncertainty quantification. Used the `self$state` [slot](https://github.com/bblodfon/mlr3extralearners/blob/main/R/learner_BART_surv_bart.R#L272) for this again. I added an example on how to use the full posterior survival matrix.
- Added a parameter to choose the [type of importance](https://github.com/bblodfon/mlr3extralearners/blob/main/R/learner_BART_surv_bart.R#L181) calculated, no design problems I guess with that?

### CODE REVIEW

**Comment by bblodfon** [comment]
@sebffischer to check/discuss during the workshop?

File: R/learner_BART_surv_bart.R [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+#' The full posterior estimates are currently stored in the
+#' `learner$state$surv_test` slot, along with the number of test observations
+#' `N`, number of unique times in the train set `K` and number of posterior
+#' draws `M`.
+#' See example for more details.
+#'
+#' Calls [BART::mc.surv.bart()] from \CRANpkg{BART}.
+#' Based on BART version 2.9.4 (2023-03-25).
+#'
+#' @section Custom mlr3 defaults:
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:20]: This section should be called "Initial parameter values" (it was wrong in the tutorial, I already opened a PR to fix this).

File: R/learner_BART_surv_bart.R [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+#' - `mc.cores` (in general use as many as possible if no issues arise):
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:21]:
```suggestion
#' - `mc.cores` is initialized to 1 to avoid threading conflicts with \CRANpkg{future}.
```

File: R/learner_BART_surv_bart.R [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+#'   - Actual default: 2
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:22]:
```suggestion
```

File: R/learner_BART_surv_bart.R [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+#'   - Adjusted value: 1
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:23]:
```suggestion
```

File: R/learner_BART_surv_bart.R [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+#'   - Reason for change: May conflict with parallelization via \CRANpkg{future}.
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:24]:
```suggestion
```

File: DESCRIPTION [hunk 6]
```diff
@@ -36,19 +36,22 @@ Suggests:
...
     dbarts,
+    dplyr,
     distr6,
     earth,
     flexsurv,
     FNN,
     formattable,
     future,
     gbm,
+    ggplot2,
```
Comment (sebffischer) [inline DESCRIPTION:54]: we should avoid depending on ggplot2, dplyr and tidyr

File: R/learner_BART_surv_bart.R [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+#' @section Initial parameter values:
+#' - `quiet`
+#'   - Default is `TRUE`, to suppress messages generated by the wrapped C++ code
+#'   during prediction.
+#'
+#' @templateVar id surv.bart
+#' @template learner
+#'
+#' @references
+#' `r format_bib("sparapani2021nonparametric", "chipman2010bart")`
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:35]: If you want to keep this example with ggplot2 etc. we could make a gallery post (https://mlr-org.com/gallery.html) out of it and link to the gallery post but I would avoid depending on tidyverse just for an example.

File: R/learner_BART_surv_bart.R [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+#' # Test
+#' p = learner$predict(task, row_ids = part$test)
+#' p$score() # C-index
+#'
+#' # Mean survival probabilities for the first 3 patients at given time points
+#' p$distr$survival(times = c(1,50,150))[,1:3]
+#'
+#' # number of posterior draws
+#' M = learner$state$M
+#' stopifnot(M == 20)
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:67]: why is this included here?

File: R/learner_BART_surv_bart.R [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+#' # number of test observations
+#' N = learner$state$N
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:69]: don't write things into the state

File: R/learner_BART_surv_bart.R [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+#' stopifnot(N == length(part$test))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:70]: why is this here?

File: R/learner_BART_surv_bart.R [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+#' # number of unique time points in the train set
+#' K = learner$state$K
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:72]: don't write things in the state (except the model)

File: R/learner_BART_surv_bart.R [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+#' stopifnot(K == length(task$unique_times(rows = part$train)))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:73]: why is this here?

File: R/learner_BART_surv_bart.R [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+      #' 2. The observed count of each feature in the trees, extracted from the
+      #' slot `varcount.mean`.
+      #' This is the default importance scores.
+      #'
+      #' In both cases, higher values signify more important variables.
+      #'
+      #' @param type Can be either `count` or `prob`.
+      #' @return Named `numeric()`.
+      importance = function(type = 'count') {
+        if (!type %in% c('prob', 'count')) stopf("type can be only 'prob' or 'count'")
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:183]:
```suggestion
assert_choice(type, c("prob", "count"))
```

File: R/learner_BART_surv_bart.R [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+      #' Recommended to use this option when `sparse = TRUE`.
+      #' 2. The observed count of each feature in the trees, extracted from the
+      #' slot `varcount.mean`.
+      #' This is the default importance scores.
+      #'
+      #' In both cases, higher values signify more important variables.
+      #'
+      #' @param type Can be either `count` or `prob`.
+      #' @return Named `numeric()`.
+      importance = function(type = 'count') {
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:182]: single 
```suggestion
      importance = function(type = "count") {
```

File: R/learner_BART_surv_bart.R [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+          sort(self$model$varcount.mean[-1], decreasing = T)
+        }
+      }
+    ),
+
+    private = list(
+      .train = function(task) {
+        pars = self$param_set$get_values(tags = "train")
+
+        x_train = as.data.frame(task$data(cols = task$feature_names))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:201]: a data table is also a data frame. Do we need this conversion?

File: R/learner_BART_surv_bart.R [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        times   = task$truth()[,1]
+        delta   = task$truth()[,2] # delta => status
+
+        # need these for predict
+        self$state$train_data = list(x_train = x_train, times = times, delta = delta)
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:206]: the `.train()` function should return a list that contains the model and this data as well. otoh storing the x_train costs a lot of memory for large datasets so it should be possible to switch it off

File: R/learner_BART_surv_bart.R [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+          .args = pars
+        )
+      },
+
+      .predict = function(task) {
+        # get parameters with tag "predict"
+        pars = self$param_set$get_values(tags = "predict")
+
+        # get newdata and ensure same ordering in train and predict
+        x_test = as.data.frame(ordered_features(task, self))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:222]: why the data.frame conversion?

File: R/learner_BART_surv_bart.R [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+#' Calls [BART::mc.surv.bart()] from \CRANpkg{BART}.
+#' Based on BART version 2.9.4 (2023-03-25).
+#'
+#' @section Custom mlr3 defaults:
+#' - `mc.cores` (in general use as many as possible if no issues arise):
+#'   - Actual default: 2
+#'   - Adjusted value: 1
+#'   - Reason for change: May conflict with parallelization via \CRANpkg{future}.
+#'
+#' @section Initial parameter values:
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:26]:
```suggestion
#' @section Custom mlr3 parameter:
```

File: R/learner_BART_surv_bart.R [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+#' - `quiet`
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:27]:
```suggestion
#' - `quiet` allows to suppress messages generated by the wrapped C++ code. Is `TRUE` by default.
```

File: R/learner_BART_surv_bart.R [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+#'   - Default is `TRUE`, to suppress messages generated by the wrapped C++ code
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:28]:
```suggestion
```

File: R/learner_BART_surv_bart.R [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+#'   during prediction.
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:29]:
```suggestion
```

File: tests/testthat/test_BART_surv_bart.R [hunk 22]
```diff
@@ -0,0 +1,8 @@
+test_that("autotest", {
+  learner = lrn("surv.bart", nskip = 1, ndpost = 1, keepevery = 1)
+  expect_learner(learner)
+  # `feat_all` task has a factor which is split to two variables which give two
```
Comment (sebffischer) [inline tests/testthat/test_BART_surv_bart.R:4]: I think this is problematic. The `$importance()` method should always return one score for each feature.

Comment (sebffischer) [reply tests/testthat/test_BART_surv_bart.R:4]: at least it should give a good error message when a factor variable was present

File: vignettes/extending.Rmd [hunk 23]
```diff
@@ -316,7 +316,7 @@ test_that("autotest", {
   learner = LearnerRegrRpartSimple$new()
 
   # basic learner properties
-  expect_learner(learner, check_man = FALSE)
```
Comment (sebffischer) [inline vignettes/extending.Rmd:319]: I think this must be kept as is as otherwise the check fails (if the learner is not installed).
When adding a learner in a package it should be `TRUE`, maybe add this as a sentence if it is not present already

File: tests/testthat/test_BART_surv_bart.R (hunk 22 above)
Comment (bblodfon) [reply tests/testthat/test_BART_surv_bart.R]: decided to remove the factor type variables. Users can encode them easily before.

File: R/learner_BART_surv_bart.R (hunk 15 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:198]: Yes sadly. In their code they check the class as `X.class = class(X)[1]` and even though a `data.table` should just play, it doesn't due to have a vector class such as: `"data.table" "data.frame"`.

File: R/learner_BART_surv_bart.R (hunk 17 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:219]: same readon as above!

File: R/learner_BART_surv_bart.R (hunk 18 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:22]: I added after this commit so it's okay now.

File: DESCRIPTION (hunk 6 above)
Comment (bblodfon) [reply DESCRIPTION:54]: removed!

**Comment by bblodfon** [comment]
Depends on https://github.com/mlr-org/mlr3proba/pull/330

**Comment by bblodfon** [comment]
@sebffischer we are ready to merge! The RCMD checks fail due to recent updates in [aorsf](https://github.com/ropensci/aorsf), I guess @bcjaeger you plan to update the learner in another PR?

**Comment by bcjaeger** [comment]
Sure, I will check it out now.

File: tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R [hunk 24]
```diff
@@ -9,7 +9,22 @@ test_that("paramtest surv.dnnsurv train", {
     "time_variable", # handled internally
     "status_variable", # handled internally
     "x", # unused
-    "y" # unused
+    "y", # unused
+    "schedule_decay", # not used
```
Comment (sebffischer) [inline tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R:13]: why did you exclude these tests?

File: R/learner_BART_surv_bart.R [hunk 25]
```diff
@@ -0,0 +1,212 @@
...
+#' @templateVar id surv.bart
+#' @template learner
+#'
+#' @references
+#' `r format_bib("sparapani2021nonparametric", "chipman2010bart")`
+#'
+#' @template seealso_learner
+#' @template example
+#' @export
+delayedAssign(
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:36]: the delayedAssign is actually not necesary (I am currently removing it for other learners here: https://github.com/mlr-org/mlr3extralearners/pull/301)

File: R/learner_BART_surv_bart.R [hunk 26]
```diff
@@ -0,0 +1,212 @@
...
+#' by default used for the calculation of survival measures that require a `distr`
+#' prediction, see more info on [PredictionSurv][mlr3proba::PredictionSurv].
+#'
+#' @section Custom mlr3 defaults:
+#' - `mc.cores` is initialized to 1 to avoid threading conflicts with \CRANpkg{future}.
+#'
+#' @section Custom mlr3 parameters:
+#' - `quiet` allows to suppress messages generated by the wrapped C++ code. Is `TRUE` by default.
+#' - `importance` allows to choose the type of importance. Default is `count`, see `importance()` for more details.
+#' - `which.curve` allows to choose which posterior draw will be used for the calculation of the `crank` prediction. By default the **median posterior** is used.
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:25]: explain how this paraeter works

File: R/learner_BART_surv_bart.R [hunk 27]
```diff
@@ -0,0 +1,212 @@
...
+#' `which.curve` decides which posterior draw (3rd dimension) will be used for the
+#' calculation of the expected mortality. Note that the median posterior is
+#' by default used for the calculation of survival measures that require a `distr`
+#' prediction, see more info on [PredictionSurv][mlr3proba::PredictionSurv].
+#'
+#' @section Custom mlr3 defaults:
+#' - `mc.cores` is initialized to 1 to avoid threading conflicts with \CRANpkg{future}.
+#'
+#' @section Custom mlr3 parameters:
+#' - `quiet` allows to suppress messages generated by the wrapped C++ code. Is `TRUE` by default.
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:23]:
```suggestion
#' - `quiet` allows to suppress messages generated by the wrapped C++ code. Is initialized to `TRUE`.
```

File: R/learner_BART_surv_bart.R [hunk 28]
```diff
@@ -0,0 +1,212 @@
...
+      #'
+      #' In both cases, higher values signify more important variables.
+      #'
+      #' @return Named `numeric()`.
+      importance = function() {
+        if (is.null(self$model$model)) {
+          stopf("No model stored")
+        }
+
+        pars = self$param_set$get_values(tags = 'train')
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:113]: double quotes and not single quotes everywhere

File: R/learner_BART_surv_bart.R [hunk 29]
```diff
@@ -0,0 +1,212 @@
...
+
+        if (pars$importance == 'prob') {
+          sort(self$model$model$varprob.mean[-1], decreasing = T)
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:116]: use `TRUE` and not `T`

File: R/learner_BART_surv_bart.R [hunk 30]
```diff
@@ -0,0 +1,212 @@
...
+          predict_types = c("crank", "distr"),
+          param_set = param_set,
+          properties = c("importance", "missings"),
+          man = "mlr3extralearners::mlr_learners_surv.bart",
+          label = "Bayesian Additive Regression Trees"
+        )
+      },
+
+      #' @description
+      #' Two types of importance scores are supported based on the initial value
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:95]:
```suggestion
      #' Two types of importance scores are supported based on the value
```

File: R/learner_BART_surv_bart.R [hunk 31]
```diff
@@ -0,0 +1,212 @@
...
+      #' for the parameter `importance`:
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:96]:
```suggestion
      #' of the parameter `importance`:
```

File: R/learner_BART_surv_bart.R [hunk 32]
```diff
@@ -0,0 +1,212 @@
...
+      #' 1. `prob`: The mean selection probability of each feature in the trees,
+      #' extracted from the slot `varprob.mean`.
+      #' If `sparse = FALSE` (default), this is a fixed constant.
+      #' Recommended to use this option when `sparse = TRUE`.
+      #' 2. `count`: The observed count of each feature in the trees, extracted
+      #' from the slot `varcount.mean`.
+      #' This is the default importance scores.
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:103]: Is it really the observed count? (I am asking because the variable is called `varcount.mean`.)

File: R/learner_BART_surv_bart.R [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+        }
+      }
+    ),
+
+    private = list(
+      .train = function(task) {
+        pars = self$param_set$get_values(tags = "train")
+        pars$importance = NULL # not used in the train function
+
+        x.train = as.data.frame(task$data(cols = task$feature_names))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:128]: in the mlr3 style guide we use `x_train` and not `x.train`

Comment (sebffischer) [reply R/learner_BART_surv_bart.R:128]: Also, is it necessary to convert to a data.frame?

File: R/learner_BART_surv_bart.R [hunk 34]
```diff
@@ -0,0 +1,212 @@
...
+        times   = task$truth()[,1]
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:129]: here you are calling `$truth()` twice, avoid this

File: R/learner_BART_surv_bart.R [hunk 35]
```diff
@@ -0,0 +1,212 @@
...
+          delta = delta
+        )
+      },
+
+      .predict = function(task) {
+        # get parameters with tag "predict"
+        pars = self$param_set$get_values(tags = "predict")
+
+        # get newdata and ensure same ordering in train and predict
+        x.test = as.data.frame(ordered_features(task, self))
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:152]: is the data.frame conversion needed?

File: R/learner_BART_surv_bart.R [hunk 36]
```diff
@@ -0,0 +1,212 @@
...
+
+        # subset parameters to use in `surv.pre.bart`
+        pars_pre = pars[names(pars) %in% c('K', 'events', 'ztimes', 'zdelta')]
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:155]: double quotes

File: R/learner_BART_surv_bart.R [hunk 37]
```diff
@@ -0,0 +1,212 @@
...
+        # See page 34-35 in Sparapani (2021) for more details
+        surv.array = aperm(
+          array(pred$surv.test, dim = c(M, K, N), dimnames = list(NULL, times, NULL)),
+          c(3, 2, 1)
+        )
+
+        # distr => 3d survival array
+        # crank => expected mortality
+        mlr3proba::.surv_return(times = times, surv = surv.array,
+          which.curve = pars$which.curve)
```
Comment (sebffischer) [inline R/learner_BART_surv_bart.R:206]: what happens if `which.curve` is `NULL` ?

Comment (sebffischer) [reply R/learner_BART_surv_bart.R:206]: to avoid this you can add the `"required"` tag to the parameter

**Review by sebffischer — CHANGES_REQUESTED** [review]

File: R/learner_BART_surv_bart.R [hunk 38]
```diff
@@ -0,0 +1,212 @@
...
+#' calculation of the expected mortality. Note that the median posterior is
+#' by default used for the calculation of survival measures that require a `distr`
+#' prediction, see more info on [PredictionSurv][mlr3proba::PredictionSurv].
+#'
+#' @section Custom mlr3 defaults:
+#' - `mc.cores` is initialized to 1 to avoid threading conflicts with \CRANpkg{future}.
+#'
+#' @section Custom mlr3 parameters:
+#' - `quiet` allows to suppress messages generated by the wrapped C++ code. Is `TRUE` by default.
+#' - `importance` allows to choose the type of importance. Default is `count`, see `importance()` for more details.
```
[the remark this answers is not in the document]
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:24]: to the `learner$importance()` slot - do you have a better way to link to it?

File: tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R (hunk 24 above)
Comment (bblodfon) [reply tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R:13]: I think there was an update in `survivalmodels` and I was getting an error at the time so I updated with the new arguments that are now used.

File: R/learner_BART_surv_bart.R (hunk 25 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:36]: Ah okay and I wanted to ask you about this! What is it? why there was a need to use it at some point? Let me know if you want it removed here or you will do it in the other PR

Comment (sebffischer) [reply R/learner_BART_surv_bart.R:36]: remove it here please. 

I think this never had the intended effect so explaining it would not be super useful :) (this was a misunderstanding from my side)

File: R/learner_BART_surv_bart.R (hunk 38 above)
[the remark this answers is not in the document]
Comment (sebffischer) [reply R/learner_BART_surv_bart.R:24]: maybe phrase it like "see documentation of method `$importance()` for more details :)

File: tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R (hunk 24 above)
Comment (sebffischer) [reply tests/testthat/test_paramtest_survivalmodels_surv_dnnsurv.R:13]: ok. 

So when parameters are outdated we should implement them for the learner and not simply ignore them in the parameter test. In the CI it seems to work, so please remove these ignored parameters again. (Once the issue arises in the CI I will fix it in a separate PR)

File: R/learner_BART_surv_bart.R (hunk 37 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:206]: no need, `NULL` is the default value and is the median survival probabilities, see [here](https://github.com/mlr-org/mlr3proba/blob/main/R/surv_return.R#L89), I have tested it also.

File: R/learner_BART_surv_bart.R (hunk 35 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:152]: Sadly yes, otherwise you get the error:
```
Error in bartModelMatrix(x.train, numcut, usequants = usequants, cont = cont,  : 
  Expecting either a factor, a vector, a matrix or a data.frame
```

File: R/learner_BART_surv_bart.R (hunk 37 above)
Comment (sebffischer) [reply R/learner_BART_surv_bart.R:206]: perfect! :)

File: R/learner_BART_surv_bart.R (hunk 32 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R:103]: The `mean`, I also rephrased it, thanks!

File: R/learner_BART_surv_bart.R (hunk 26 above)
Comment (bblodfon) [reply R/learner_BART_surv_bart.R]: solved now

**Review by sebffischer — APPROVED** [review]
